README.md

# php-psr

[![GitHub Linux Build Status](https://github.com/jbboehr/php-psr/workflows/linux/badge.svg)](https://github.com/jbboehr/php-psr/actions?query=workflow%3Alinux)
[![GitHub OSX Build Status](https://github.com/jbboehr/php-psr/workflows/osx/badge.svg)](https://github.com/jbboehr/php-psr/actions?query=workflow%3Aosx)
[![GitHub Windows Build Status](https://github.com/jbboehr/php-psr/workflows/windows/badge.svg)](https://github.com/jbboehr/php-psr/actions?query=workflow%3Awindows)
[![GitHub Docker Build Status](https://github.com/jbboehr/php-psr/workflows/docker/badge.svg)](https://github.com/jbboehr/php-psr/actions?query=workflow%3Adocker)
[![Appveyor Build Status][:badge-appveyor:]][:build-appveyor:]
[![Coverage Status][:badge-coveralls:]][:build-coveralls:]
[![License][:badge-license:]][:ext-license:]

This PHP extension provides the interfaces from the [PSR][:psr-fig:] standards as established by the [PHP-FIG][:php-fig:] group.
You can use interfaces provided by this extension in another extension easily - see this [example][:example:].

As of v1.2 of the extension, classes are defined in the `PsrExt` namespace and aliased into the `Psr` namespace.

## Interfaces

| PSR | Reference | Tested against |
| --- | --- | --- |
| [PSR-3] | [psr/log] [*](#psrlogunimpl) | [monolog/monolog] |
| [PSR-6] | [psr/cache] | [psx/cache], [php-fig/cache-util] |
| [PSR-7] | [psr/http-message] | [laminas/laminas-diactoros] |
| [PSR-11] | [psr/container] | [psx/dependency] |
| [PSR-13] | [psr/link] | [php-fig/link-util] |
| [PSR-14] | [psr/event-dispatcher] | [crell/tukio], [php-fig/event-dispatcher-util] |
| [PSR-15] | [psr/http-server-handler] | [equip/dispatch] |
| [PSR-15] | [psr/http-server-middleware] | [middlewares/request-handler] |
| [PSR-16] | [psr/simple-cache] | [psx/cache] |
| [PSR-17] | [psr/http-factory] | [http-interop/http-factory-guzzle] |
| [PSR-18] | [psr/http-client] | [ricardofiorani/guzzle-psr18-adapter] |

## Installation

### Linux / macOS

Prerequisite packages are:

- PHP development headers and tools, `php` >= 7.3
- `gcc` >= 4.4 | `clang` >= 3.x | `vc` >= 11
- GNU `make` >= 3.81
- `automake`
- `autoconf`

You will need the PHP development headers. If PHP was manually installed, these should be available by default. Otherwise, you will need to fetch them from a repository.

```bash
git clone https://github.com/jbboehr/php-psr.git
cd php-psr
phpize
./configure
make
make test
sudo make install
```

If you have specific PHP versions running:

```bash
git clone https://github.com/jbboehr/php-psr.git
cd php-psr
/usr/local/bin/phpize
./configure --with-php-config=/usr/local/bin/php-config
make
make test
sudo make install
```

Add the extension to your *php.ini*:

```ini
echo extension=psr.so | tee -a /path/to/your/php.ini
```

Finally, _restart the web server_.

### PECL / Windows

You may also be able to install this extension via [PECL][:pecl-psr:]:

```bash
pecl install psr
```

or by downloading a DLL from [PECL][:pecl-psr:] or [windows.php.net][:windows-psr:] and placing it in the appropriate directory.

### Nix / NixOS

```bash
nix-env -i -f https://github.com/jbboehr/php-psr/archive/master.tar.gz
```

with a custom version of PHP:

```bash
nix-env -i -f https://github.com/jbboehr/php-psr/archive/master.tar.gz --arg php '(import <nixpkgs> {}).php71'
```

or, in a `.nix` file:

```nix
(import <nixpkgs> {}).callPackage (import (fetchTarball {
  url = https://github.com/jbboehr/php-psr/archive/v1.2.0.tar.gz;
  sha256 = "1c4jc09d46ac43kkckil8l76is973czwp0g6vhpnv5105l7z3zgi";
})) {}
```

### Using with composer

In your project, you can prevent the installation of the unnecessary composer packages provided by this extension by adding the following to your `composer.json`. You will need to make sure the extension is installed and enabled in your PHP configuration. You may also want to consider leaving them installed to provide stubs to your IDE. This configuration is not suitable for a library published to packagist.

<a name="psrlogunimpl">Note:</a> `LoggerInterfaceTest` and `TestLogger` from `psr/log` are not implemented. If you use these classes, you should keep `psr/log` installed ([discussion](https://github.com/jbboehr/php-psr/issues/57)).

```json
{
    "name": "sample/app",
    "require": {
        "ext-psr": "*"
    },
    "provide": {
        "psr/log": "1.0.0",
        "psr/cache": "1.0.0",
        "psr/http-message": "1.0.0",
        "psr/container": "1.1.1",
        "psr/link": "1.0.0",
        "psr/event-dispatcher": "1.0.0",
        "psr/http-server-handler": "1.0.0",
        "psr/http-server-middleware": "1.0.0",
        "psr/simple-cache": "1.0.0",
        "psr/http-factory": "1.0.0",
        "psr/http-client": "1.0.0"
    },
}
```

This will use this PHP extension instead of downloading from packagist.

## Credits

* [@sergeyklay][:github-sergeyklay:] - Appveyor and documentation improvements
* [@flyinghail][:github-flyinghail:] - PSR-14, PSR-15, PSR-17, and PSR-18 support
* [@Jan-E][:github-jan-e:] - Original appveyor template
* [All Contributors](https://github.com/jbboehr/php-psr/graphs/contributors)

## License

This project is open source software licensed under the Simplified BSD License.
See the [LICENSE.md][:ext-license:] file for more information.

PSR Interfaces: Copyright (c) 2012-present [PHP Framework Interoperability Group][:php-fig:].

[:psr-fig:]: https://www.php-fig.org/psr
[:php-fig:]: https://www.php-fig.org
[:pecl-psr:]: https://pecl.php.net/package/psr
[:badge-appveyor:]: https://ci.appveyor.com/api/projects/status/x1ymkqggy1mkl0ux/branch/master?svg=true
[:badge-coveralls:]: https://coveralls.io/repos/jbboehr/php-psr/badge.svg?branch=master&service=github
[:badge-license:]: https://img.shields.io/badge/license-BSD-brightgreen.svg
[:build-appveyor:]: https://ci.appveyor.com/project/jbboehr/php-psr/branch/master
[:build-coveralls:]: https://coveralls.io/github/jbboehr/php-psr?branch=master
[:ext-license:]: https://github.com/jbboehr/php-psr/blob/master/LICENSE.md
[:example:]: https://github.com/jbboehr/php-handlebars/blob/v0.7.1/impl.c#L213-L215
[:windows-psr:]: http://windows.php.net/downloads/pecl/releases/psr/

[PSR-3]: http://www.php-fig.org/psr/psr-3
[PSR-6]: https://www.php-fig.org/psr/psr-6
[PSR-7]: https://www.php-fig.org/psr/psr-7
[PSR-11]: https://www.php-fig.org/psr/psr-11
[PSR-13]: https://www.php-fig.org/psr/psr-13
[PSR-14]: http://www.php-fig.org/psr/psr-14
[PSR-15]: http://www.php-fig.org/psr/psr-15
[PSR-16]: https://www.php-fig.org/psr/psr-16
[PSR-17]: https://www.php-fig.org/psr/psr-17
[PSR-18]: https://www.php-fig.org/psr/psr-18

[psr/log]: https://github.com/php-fig/log
[psr/cache]: https://github.com/php-fig/cache
[psr/http-message]: https://github.com/php-fig/http-message
[psr/container]: https://github.com/php-fig/container
[psr/link]: https://github.com/php-fig/link
[psr/simple-cache]: https://github.com/php-fig/simple-cache
[psr/http-server-handler]: https://github.com/php-fig/http-server-handler
[psr/http-server-middleware]: https://github.com/php-fig/http-server-middleware
[psr/http-factory]: https://github.com/php-fig/http-factory
[psr/http-client]: https://github.com/php-fig/http-client
[psr/event-dispatcher]: https://github.com/php-fig/event-dispatcher

[equip/dispatch]: https://github.com/equip/dispatch
[middlewares/request-handler]: https://github.com/middlewares/request-handler
[http-interop/http-factory-guzzle]: https://github.com/http-interop/http-factory-guzzle
[ricardofiorani/guzzle-psr18-adapter]: https://github.com/ricardofiorani/guzzle-psr18-adapter
[monolog/monolog]: https://github.com/Seldaek/monolog
[psx/cache]: https://github.com/apioo/psx-cache
[psx/dependency]: https://github.com/apioo/psx-dependency
[php-fig/link-util]: https://github.com/php-fig/link-util
[crell/tukio]: https://github.com/Crell/Tukio.git
[php-fig/event-dispatcher-util]: https://github.com/php-fig/event-dispatcher-util
[php-fig/cache-util]: https://github.com/php-fig/cache-util
[laminas/laminas-diactoros]: https://github.com/laminas/laminas-diactoros
[php-fig/event-dispatcher-util]: https://github.com/php-fig/event-dispatcher-util

[:github-sergeyklay:]: https://github.com/sergeyklay
[:github-jan-e:]: https://github.com/Jan-E
[:github-flyinghail:]: https://github.com/flyinghail


CHANGELOG.md
# Changelog

All Notable changes to `php-psr` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## [1.2.0] - 2021-12-11

### Added
- PHP 8.1 support (@Jeckerson)

### Changed
- All classes are now defined in a separate namespace `PsrExt` and aliased into the `Psr` 
namespace to aid in debuggging compatibility issues for framework maintainers. ([GH-88](https://github.com/jbboehr/php-psr/issues/88))

## [1.1.0] - 2021-04-10

### Changed
- Update psr/container to 1.1.1

### Added
- Version ID constants to each header

### Fixed
- Use __dllspec(dllimport) where appropriate in php_psr.h on windows

## [1.0.1] - 2020-11-12

### Fixed 
- Test failures on PHP 8 in psr-http-client and psr-http-message due to changes in inheritance
- Compile failures with clang
- Compatibility with API changes to PHP 8 in psr-log (@remicollet)
- Report proper default value in reflection for LoggerInterface - PHP 8 only ([GH-78](https://github.com/jbboehr/php-psr/issues/78))

## [1.0.0] - 2020-02-18

### Changed
- Drop max PHP version constraint in package.xml

## [0.7.0] - 2019-08-12

### Added
- Preliminary PHP 7.4 and 8.0 (master) support
- PSR-14 Event Dispatcher support

### Removed
- PHP 5 support

## [0.6.1] - 2018-11-13

### Fixed 
- Test failures on PHP 5.6

## [0.6.0] - 2018-11-11

### Added
- PSR-18 HTTP Client support

## [0.5.1] - 2018-10-29

### Fixed 
- `Psr\Http\Message\ServerRequestInterface` did not actually extend `Psr\Http\Message\RequestInterface`

## [0.5.0] - 2018-09-10

### Added
- PSR-17 HTTP Factories support
- Whitelist PHP 7.3

## [0.4.0] - 2018-03-05

### Added
- PSR-15 HTTP Handlers support

## [0.3.0] - 2018-02-08

### Added
- PSR-11 Container support
- PSR-13 Link support
- PSR-16 Simple Cache support
- Windows support

### Removed
- No longer run CI against PHP 5 due to lack of support in libraries

### Fixed 
- TypeError when no argument is passed to methods in AbstractLogger

## [0.2.3] - 2016-05-05

### Changed
- Changed the license from `LGPLv3` to `Simplified BSD License`
- Move PHP5/7 compatibility macros into separate header

## [0.2.2] - 2016-03-16

### Added
- arginfo for interfaces in headers

### Fixed 
- Memory leak in Psr\Log\AbstractLogger::log()
 
## [0.2.1] - 2015-12-15

### Changed
- PSR-6 [psr/cache](https://github.com/php-fig/cache) v1.0.0 is released without changes to interfaces since commit [4908941](https://github.com/php-fig/cache/commit/4908941)
- Test PSR-6 interfaces against Stash v1.0.0-dev branch

### Fixed
- Memory leak in Psr\Log\AbstractLogger::log() with PHP7

## [0.2.0] - 2015-12-09

### Changed
- Update PSR-6 interfaces for commits up to [4908941](https://github.com/php-fig/cache/commit/4908941)

### Fixed
- Invalid read in Psr\Log\AbstractLogger::log()

## 0.1.0 - 2015-08-07

### Added
- PSR-3 [psr/log](https://github.com/php-fig/log) v1.0.0 interfaces and classes
- PSR-6 [commit 2b812a4](https://github.com/php-fig/fig-standards/commit/2b812a4) interfaces
- PSR-7 [psr/http-message](https://github.com/php-fig/http-message) v1.0.0 interfaces


[Unreleased]: https://github.com/jbboehr/php-psr/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/jbboehr/php-psr/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/jbboehr/php-psr/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/jbboehr/php-psr/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/jbboehr/php-psr/compare/v0.7.0...v1.0.0
[0.7.0]: https://github.com/jbboehr/php-psr/compare/v0.6.1...v0.7.0
[0.6.1]: https://github.com/jbboehr/php-psr/compare/v0.6.0...v0.6.1
[0.6.0]: https://github.com/jbboehr/php-psr/compare/v0.5.1...v0.6.0
[0.5.1]: https://github.com/jbboehr/php-psr/compare/v0.5.0...v0.5.1
[0.5.0]: https://github.com/jbboehr/php-psr/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/jbboehr/php-psr/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/jbboehr/php-psr/compare/v0.2.3...v0.3.0
[0.2.3]: https://github.com/jbboehr/php-psr/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/jbboehr/php-psr/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/jbboehr/php-psr/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/jbboehr/php-psr/compare/v0.1.0...v0.2.0


php_psr.c

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"
#include "php_ini.h"
#include "ext/standard/info.h"
#include "ext/spl/spl_exceptions.h"
#include "zend_API.h"
#include "zend_interfaces.h"

#include "php_psr.h"
#include "psr_cache.h"
#include "psr_container.h"
#include "psr_http_message.h"
#include "psr_link.h"
#include "psr_log.h"
#include "psr_simple_cache.h"
#include "psr_http_server_handler.h"
#include "psr_http_server_middleware.h"
#include "psr_http_factory.h"
#include "psr_http_client.h"
#include "psr_event_dispatcher.h"

static PHP_MINIT_FUNCTION(psr)
{
    PHP_MINIT(psr_cache)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_container)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_http_message)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_link)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_log)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_simple_cache)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_http_server_handler)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_http_server_middleware)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_http_factory)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_http_client)(INIT_FUNC_ARGS_PASSTHRU);
    PHP_MINIT(psr_event_dispatcher)(INIT_FUNC_ARGS_PASSTHRU);

    return SUCCESS;
}

static PHP_MINFO_FUNCTION(psr)
{
    php_info_print_table_start();
    php_info_print_table_row(2, "Version", PHP_PSR_VERSION);
    php_info_print_table_row(2, "Released", PHP_PSR_RELEASE);
    php_info_print_table_row(2, "Authors", PHP_PSR_AUTHORS);
    php_info_print_table_row(2, "PSR-3 Log Version", PHP_PSR_LOG_VERSION);
    php_info_print_table_row(2, "PSR-6 Cache Version", PHP_PSR_CACHE_VERSION);
    php_info_print_table_row(2, "PSR-7 Http Message Version", PHP_PSR_HTTP_MESSAGE_VERSION);
    php_info_print_table_row(2, "PSR-11 Container Version", PHP_PSR_CONTAINER_VERSION);
    php_info_print_table_row(2, "PSR-13 Link Version", PHP_PSR_LOG_VERSION);
    php_info_print_table_row(2, "PSR-14 Event Dispatcher", PHP_PSR_EVENT_DISPATCHER_VERSION);
    php_info_print_table_row(2, "PSR-15 HTTP Handlers (Server Handler)", PHP_PSR_HTTP_SERVER_HANDLER_VERSION);
    php_info_print_table_row(2, "PSR-15 HTTP Handlers (Middleware)", PHP_PSR_HTTP_SERVER_MIDDLEWARE_VERSION);
    php_info_print_table_row(2, "PSR-16 Simple Cache Version", PHP_PSR_SIMPLE_CACHE_VERSION);
    php_info_print_table_row(2, "PSR-17 HTTP Factories", PHP_PSR_HTTP_FACTORY_VERSION);
    php_info_print_table_row(2, "PSR-18 HTTP Client", PHP_PSR_HTTP_CLIENT_VERSION);
    php_info_print_table_end();
}

static const zend_module_dep psr_deps[] = {
    ZEND_MOD_REQUIRED("spl")
    ZEND_MOD_END
};

zend_module_entry psr_module_entry = {
    STANDARD_MODULE_HEADER_EX,
    NULL,
    psr_deps,
    PHP_PSR_NAME,                       /* Name */
    NULL,                               /* Functions */
    PHP_MINIT(psr),                     /* MINIT */
    NULL,                               /* MSHUTDOWN */
    NULL,                               /* RINIT */
    NULL,                               /* RSHUTDOWN */
    PHP_MINFO(psr),                     /* MINFO */
    PHP_PSR_VERSION,                    /* Version */
    STANDARD_MODULE_PROPERTIES
};

#ifdef COMPILE_DL_PSR
    ZEND_GET_MODULE(psr)      // Common for all PHP extensions which are build as shared modules
#endif

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: fdm=marker
 * vim: et sw=4 ts=4
 */
